### PR TITLE
First draft of mqtt as a bus service

### PR DESCRIPTION
**Pull request by gdraynz**
## Run
Configuration to enable mqtt :
```
{
    "bus": {
        "service": "mqtt",       (default: xmpp)
        "name": "testnyuki",     (required to identify the nyuki)
        "scheme": "ws",
        "host": "localhost"
    }
}
```
Start the mqtt broker using :
```bash
docker pull surycat/mosquitto:unsecure
docker run -it -p 1883:1883 surycat/mosquitto:unsecure
```
## Usage
Listen to a mqtt pattern (starts with `publications/` and anything after)
```python
async def handle_pub(topic, data):
    print(topic)
    print(data)

await self.bus.subscribe('publications/#', handle_pub)
```
Publish to a mqtt topic
```python
# Implies topic='publications/nyuki_name'
await self.bus.publish({'something': 'here'})
await self.bus.publish({'something': 'here'}, topic='a/custom/topic')
```

### CODE REVIEW

File: nyuki/nyuki.py [hunk 1]
```diff
@@ -70,9 +69,14 @@ def __init__(self, **kwargs):
         self._services = ServiceManager(self)
         self._services.add('api', Exposer(self))
 
-        # Add bus service if in conf file
-        if self._config.get('bus') is not None:
-            self._services.add('bus', Bus(self))
+        # Add bus service if in conf file, xmpp (default) or mqtt
+        bus_config = self._config.get('bus')
+        if bus_config is not None:
+            bus_service = bus_config.get('service', 'xmpp')
```
Comment (thavel) [inline nyuki/nyuki.py]: Let's call this `protocol` instead of `service`.

File: nyuki/bus/mqtt.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+from nyuki.bus import reporting
+from nyuki.bus.persistence import BusPersistence, EventStatus, PersistenceError
+from nyuki.services import Service
+
+
+log = logging.getLogger(__name__)
+
+
+class MqttBus(Service):
```
Comment (thavel) [inline nyuki/bus/mqtt.py]: Wouldn't be clever to gather common code (xmpp and mqtt clients) into a single meta-class?

File: nyuki/bus/mqtt.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+    CONF_SCHEMA = {
+        "type": "object",
+        "required": ["bus"],
+        "properties": {
+            "bus": {
+                "type": "object",
+                "required": ["host", "name"],
+                "properties": {
+                    "certificate": {"type": "string", "minLength": 1},
```
Comment (thavel) [inline nyuki/bus/mqtt.py]: SSL requires more than juste one certificate:
* `cafile`
* `certfile`
* `keyfile`

File: nyuki/bus/mqtt.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+    BASE_PUB = 'publications'
+    BASE_RESP = 'responses'
+
+    def __init__(self, nyuki, loop=None):
+        self._nyuki = nyuki
+        self._loop = loop or asyncio.get_event_loop()
+        self._host = None
+        self._self_topic = None
+        self.client = MQTTClient(
```
Comment (thavel) [inline nyuki/bus/mqtt.py]: You should add the following info to the client's config if SSL is enabled:
```python
{
    'certfile': 'server.crt',
    'keyfile': 'server.key'
}
```

File: nyuki/bus/mqtt.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    CONF_SCHEMA = {
+        "type": "object",
+        "required": ["bus"],
+        "properties": {
+            "bus": {
+                "type": "object",
+                "required": ["host", "name"],
+                "properties": {
+                    "certificate": {"type": "string", "minLength": 1},
+                    "host": {"type": "string", "minLength": 1},
```
Comment (thavel) [inline nyuki/bus/mqtt.py]: In your PR, the protocol has to be set in the `host`. I'd rather use a `protocol` key (possible values: `mqtt` - default, `mqtts`, `ws`, `wss`).
Therefore, `cafile`, `certfile` and `keyfile` become mandatory if `wss` or `mqtts` are set.

File: nyuki/capabilities.py [hunk 6]
```diff
@@ -107,15 +107,17 @@ def _find(self, capa_name):
...
             return
 
-        # TODO: use asyncio.ensure_future when Python 3.4.4 will be released
-        future = asyncio.ensure_future(capa.wrapper(request), loop=self._loop)
-        return future
+        response = await capa.wrapper(body)
+        if response.headers.get('content-type') == 'application/json':
+            return response.body
+        else:
+            return json.dumps({'result': response.body})
```
Comment (thavel) [inline nyuki/capabilities.py]: What's the point to do this?

Comment (gdraynz) [inline nyuki/capabilities.py]: Unused residual code, will remove

File: nyuki/bus/mqtt.py (hunk 2 above)
Comment (gdraynz) [inline nyuki/bus/mqtt.py]: Thing is apart from the API methods' names, there not really any "common code"

**Comment by thavel** [comment]
To be fixed:
* Persistence publish replay for MQTT.
* Fix reporting for MQTT.

**Comment by thavel** [comment]
Back in XMPP, while publish a report, got:
```
2016-07-21 17:31:27,549  WARNING  [nyuki.bus.xmpp] Publication timed out, disconnecting slixmpp
2016-07-21 17:31:27,549  INFO     [nyuki.bus.persistence.persistence] Updating status of event '73a60d69-4552-401a-abd6-30cf8478fd71' to 'EventStatus.FAILED'
2016-07-21 17:31:27,550  WARNING  [nyuki.bus.xmpp] Disconnected from the bus
```
And it keeps connect/disconnect...

**Pull request merged**